etl::string_stream double formatting broken

I just upgraded release `20.2.2` to `20.21.0` and have noticed that writing a double to a string stream using `precision` has changed to a broken behavior.

I've narrowed it down to the following section of our code.

```c++
    WW_LOG_TRACE("value=%f", value);
    etl::string<20> data_string;
    etl::string_stream data_stream(data_string);
    etl::format_spec format;
    format.precision(10);
    data_stream << format << value;
    WW_LOG_TRACE("data_string=%s", data_string.c_str())
```

Produces the following logs

```
18: [2021-12-09 16:49:05.308] [firmware] [trace] [utility/system] __write_double:292 value=5.987000
18: [2021-12-09 16:49:05.308] [firmware] [trace] [utility/system] __write_double:304 data_string=5.139173455
```


The problem is that the value `5.987000` should be formatted to a string `5.9870000000`.

I have narrowed down the introduction of the bug to the following commit

df29374c6ea5cd48173d8779ba1c46ace4a553dc

Using the following unit test

```c++
    TEST(test_format_stream_float)
    {
      const double value = 1.2345;

      String str;
      Stream ss(str);

      Format format;
      format.precision(10);
      ss << format << value;

      CHECK_EQUAL(String(STR("1.2345000000")), str);
    }
```

```
etl/test/test_string_stream.cpp:326:1: error: Failure in test_format_stream_float: Expected 1.2345000000 but was 1.0330660338
```


Will create a PR if I can find the root cause

--- a/include/etl/private/to_string_helper.h
+++ b/include/etl/private/to_string_helper.h
@@ -56,8 +56,10 @@ namespace etl
   {
 #if ETL_NOT_USING_64BIT_TYPES
     typedef int32_t workspace_t;
+    typedef uint32_t uworkspace_t;
 #else
     typedef int64_t workspace_t;
+    typedef uint64_t uworkspace_t;
 #endif
 
     //***************************************************************************
@@ -314,13 +316,20 @@ namespace etl
         // Make sure we format the two halves correctly.
         uint32_t max_precision = etl::numeric_limits<T>::digits10;
 
+#if ETL_NOT_USING_64BIT_TYPES
+        if (max_precision > 9)
+        {
+          max_precision = 9;
+        }
+#endif
+
         etl::basic_format_spec<TIString> integral_format = format;
         integral_format.decimal().width(0).precision(format.get_precision() > max_precision ? max_precision : format.get_precision());
 
         etl::basic_format_spec<TIString> fractional_format = integral_format;
         fractional_format.width(integral_format.get_precision()).fill(type('0')).right();
 
-        uint32_t multiplier = 1U;
+        uworkspace_t multiplier = 1U;
 
         for (uint32_t i = 0U; i < fractional_format.get_precision(); ++i)
         {
@@ -329,10 +338,10 @@ namespace etl
 
         // Find the integral part of the floating point
         T f_integral = floor(etl::absolute(value));
-        uint32_t integral = static_cast<uint32_t>(f_integral);
+        uworkspace_t integral = static_cast<uworkspace_t>(f_integral);
 
         // Find the fractional part of the floating point.
-        uint32_t fractional = static_cast<uint32_t>(round((etl::absolute(value) - f_integral) * multiplier));
+        uworkspace_t fractional = static_cast<uworkspace_t>(round((etl::absolute(value) - f_integral) * multiplier));
 
         // Check for a rounding carry to the integral.
         if (fractional == multiplier)

--- a/test/test_to_string.cpp
+++ b/test/test_to_string.cpp
@@ -598,7 +598,7 @@ namespace
     }
     
     //*************************************************************************
-    TEST(test_double_regression_issue_479)
+    TEST(test_double_formatting_10_decimal_point)
     {
       etl::string<20> result_d;
       double value_d = 1.2345;

--- a/test/test_to_string.cpp
+++ b/test/test_to_string.cpp
@@ -596,6 +596,18 @@ namespace
       CHECK_EQUAL(etl::string<20>(STR("-123.4568")).c_str(), result_i.c_str());
       CHECK_EQUAL(result_d.c_str(), result_i.c_str());
     }
+    
+    //*************************************************************************
+    TEST(test_double_regression_issue_479)
+    {
+      etl::string<20> result_d;
+      double value_d = 1.2345;
+      
+      Format format = Format().precision(10);
+      etl::to_string(value_d, result_d, format);
+
+      CHECK_EQUAL(etl::string<20>(STR("1.2345000000")).c_str(), result_d.c_str());
+    }
 
     //*************************************************************************
     TEST(test_integer_denominator_larger_width)

